File: csharp/nth-prime/NthPrime.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class NthPrime
{
    public static int Prime(int nth)
    {
        if (nth < 1)
            throw new ArgumentOutOfRangeException(nameof(nth), nth, "Parameter should be a positive number");

        return Primes().Skip(nth - 1).First();
    }

    private static IEnumerable<int> Primes()
    {
        var primes = new HashSet<int>();
        
        return Enumerable
            .Range(2, int.MaxValue - 2)
            .Where(IsPrime)
            .Where(primes.Add);

        bool IsPrime(int n) => primes.All(x => n % x != 0);
    }
}

File: kotlin/nth-prime/src/main/kotlin/NthPrime.kt
object Prime {

    fun nth(n: Int): Int {
        require(n > 0) {
            "There is no zeroth prime."
        }
        
        return primes.drop(n - 1).first()
    }
    
    private val primes = sequence {
        val state = mutableListOf<Int>()
        
        for (i in 2..Int.MAX_VALUE) {
            if (state.all { i % it != 0 }) {
                yield(i)
                state += i
            }
        }
    }
}
